.ci/release/Jenkinsfile

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env groovy
+
+@Library('apm@current') _
+pipeline {
+  agent { label 'linux && immutable' }
+  environment {
+    REPO = 'apm-agent-java'
+    BASE_DIR = "src/github.com/elastic/${env.REPO}"
+    NOTIFY_TO = credentials('notify-to')
+    DOCKERHUB_SECRET = 'secret/apm-team/ci/elastic-observability-dockerhub'
+    ELASTIC_DOCKER_SECRET = 'secret/apm-team/ci/docker-registry/prod'
+    NEXUS_SECRET = 'secret/apm-team/ci/nexus'
+    MAVEN_CONFIG = '-Dmaven.repo.local=.m2'
+    OPBEANS_REPO = 'opbeans-java'
+
+  }
+  options {
+    timeout(time: 3, unit: 'HOURS')
+    buildDiscarder(logRotator(numToKeepStr: '20', artifactNumToKeepStr: '20', daysToKeepStr: '30'))
+    timestamps()
+    ansiColor('xterm')
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+
+  }
+  parameters {
+    string(name: 'MAVEN_CONFIG', defaultValue: '-B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn', description: 'Additional maven options.')
+    string(name: 'branch_specifier', defaultValue: 'master')
+  }
+
+  stages {
+    stage('Initializing'){
+      options { skipDefaultCheckout() }
+      environment {
+        HOME = "${env.WORKSPACE}"
+        JAVA_HOME = "${env.HUDSON_HOME}/.java/java10"
+        PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+        MAVEN_CONFIG = "${params.MAVEN_CONFIG} ${env.MAVEN_CONFIG}"
+      }
+      stages {
+        /**
+         Checkout the code
+        */
+        stage('Checkout') {
+          steps {
+            dir("${BASE_DIR}") {
+              // This is required to ensure that Maven is working with a real branch
+              sh(script: "git checkout -f {$branch_specifier}")
+            }
+          }
+        }
+      }
+    }
+    stage('Release') {
+      options { skipDefaultCheckout () }
+      stages{
+        // 1. Check to see if oss.sonatype.org is up. We throw an exception and fail the job if there is no response.
+        stage('Check oss.sonatype.org') {
+          steps {
+            // If this fails, an exception should be thrown and execution will halt
+            httpRequest(url: "https://oss.sonatype.org/#welcome")
+          }
+        }
+        // 2. Review project version.
+        stage('Review project version') {
+          steps {
+            script {
+              // Show the select input modal
+              def INPUT_PARAMS = input message: 'Do you wish to update the version?', ok: 'Next', parameters: [choice(name: 'UPDATE_CHOICE', choices:['Yes', 'No'], description: "This will run `mvn release:update-versions`")]
```

Comment:
Instead of updating the version here, I wonder if we should instead require the release version to be set in the release job. See https://maven.apache.org/maven-release/maven-release-plugin/examples/non-interactive-release.html

We could then also require to set the next dev version. The release plugin offers system properties for the release and next dev version: `-DreleaseVersion=1.2.0  -DdevelopmentVersion=1.3.0-SNAPSHOT`.

@eyalkoren @SylvainJuge WDYT?

Reply:
I'm not really sure to understand how this would be different: do we make the release job just "release current version", which is read from `pom.xml` ?

What if we keep having this prompt for release version and the default value is read from `pom.xml`, if a different value is set just updates version before executing release ?

Having this version set/bump done at CI level allows to avoid running any local git command, which is one less opportunity for human errors.

I'm also ok for doing an increment on the minor version number, but would rather use `mvn versions:set` and not `mvn:release` plugin. for the reason I mentioned above (which I can definitely get by).

Reply:
> but would rather use mvn versions:set and not mvn:release plugin

During a release:prepare and after a release:perform the versions get updated by the release plugin anyways.

So in the current state of this PR, we would update the version in step 2, then once again in on internal-ci via the `release:prepare` stage (it removes the `-SNAPSHOT`) and finally after the release is done, after the `release:perform` stage (increments the patch version and adds `-SNAPSHOT`, in most cases we actually want to increment the minor version).

What I'm suggesting is to modify the version 2 times instead of 3.

Reply:
Hi @felixbarny and @SylvainJuge We've let this discussion sit for a while. I don't have any strong opinions on the way we handle this, but I'd like to come to a consensus so we can get this merged. What's the best way to find a path forward here?

Reply:
don't see this as blocking the PR, that's something we can always improve later on